File: problems/70_climbing-stairs/main.cpp
#include <iostream>
#include <utility>
#include <vector>

class Solution
{
public:
    std::vector<int> dp;
    int climb(int n)
    {
        if (dp[n] != -1)
            return dp[n];
        if (n == 0 || n == 1)
            return dp[n] = 1;
        return dp[n] = climb(n - 1) + climb(n - 2);
    }
    int climbStairs(int n)
    {
        dp.assign(n + 1, -1);
        return climb(n);
    }
    int climbStairs_2(int n)
    {
        if (n == 0 || n == 1)
            return 1;
        int prev_1 = 1, prev_2 = 1, curr = 1;
        for (int i = 2; i <= n; i++)
        {
            curr = prev_1 + prev_2;
            prev_1 = prev_2;
            prev_2 = curr;
        }
        return curr;
    }
};

int main()
{
    Solution solution;
    std::vector<std::pair<int, int>> samples = {
        {2, 2},
        {3, 3},
    };

    for (const auto &[n, expected] : samples)
    {
        int result = solution.climbStairs(n);
        std::cout << "input: n = " << n
                  << ", placeholder result: " << result
                  << ", expected: " << expected << '\n';
    }

    return 0;
}

File: problems/70_climbing-stairs/main.rs
struct Solution;

impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {
        let mut dp = vec![-1; (n + 1) as usize];
        Self::climb(n, &mut dp)
    }

    fn climb(n: i32, dp: &mut Vec<i32>) -> i32 {
        if dp[n as usize] != -1 {
            return dp[n as usize];
        }
        if n == 0 || n == 1 {
            dp[n as usize] = 1;
        } else {
            dp[n as usize] = Self::climb(n - 1, dp) + Self::climb(n - 2, dp);
        }
        dp[n as usize]
    }

    pub fn climb_stairs_2(n: i32) -> i32 {
        if n == 0 || n == 1 {
            return 1;
        }
        let mut prev_1 = 1;
        let mut prev_2 = 1;
        let mut curr = 1;
        for _ in 2..=n {
            curr = prev_1 + prev_2;
            prev_1 = prev_2;
            prev_2 = curr;
        }
        curr
    }
}

fn main() {
    let samples = vec![(2, 2), (3, 3)];

    for (n, expected) in samples {
        let result = Solution::climb_stairs(n);
        println!(
            "input: n = {}, placeholder result: {}, expected: {}",
            n, result, expected
        );
    }
}
